### PR TITLE
Task2 Admin CTG Page

### PR DESCRIPTION
**Pull request by Olya27**

### CODE REVIEW

File: pages/edit_categoty.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
+from selenium import webdriver
+from selenium.webdriver.common.by import By
+
+class AdminAddCategoryPage():
+
```
Comment (TanyaKalistruk) [inline pages/edit_categoty.py:5]: Please, add docstrings to the class and methods

File: pages/edit_categoty.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    ADD_CATEGORY_BUTTON_XPATH = "//button[contains(text(),'Add element')]"
+    EDIT_CATEGORY_BUTTON_XPATH = "//table[@class='table w-100 m-auto']/tr[3]/td[4/div/button]"
```
Comment (TanyaKalistruk) [inline pages/edit_categoty.py:7]: Is it correct xpath?

File: pages/edit_categoty.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        element=self.driver.find_element(By.XPATH, self.SUBMIT_EDIT_CATEGORY_XPATH)
+        element.click()
+
+    def click_cancel_editing(self):
+        element=self.driver.find_element(By.XPATH, self.CANCEL_EDIT_CATEGORY_XPATH)
+        element.click()
+
+    def click_delete_category(self):
+        element=self.driver.find_element(By.XPATH, self.DELETE_CATEGORY_XPATH)
+        element.click()
```
Comment (TanyaKalistruk) [inline pages/edit_categoty.py:38]: Empty line in the end of file

File: pages/edit_categoty.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
+from selenium import webdriver
+from selenium.webdriver.common.by import By
+
+class AdminAddCategoryPage():
+
+    '''Locators and methjods for Admin Edit category page'''
```
Comment (vladskakun) [inline pages/edit_categoty.py:6]: Docstring in class must consist in "", not in ''.

File: pages/edit_categoty.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+    ADD_CATEGORY_BUTTON_XPATH = "//button[contains(text(),'Add element')]"
+    EDIT_CATEGORY_BUTTON_XPATH = "//table[@class='table w-100 m-auto']/tr[3]/td[4]/div/button"
+    NAME_CATEGORY_INPUT_XPATH = "//table[@class='table w-100 m-auto']/tr[3]/td[1]/input"
+    SUBMIT_EDIT_CATEGORY_XPATH = "//table[@class='table w-100 m-auto']/tr[3]/td[4]/button[@type='submit']"
+    CANCEL_EDIT_CATEGORY_XPATH = "//table[@class='table w-100 m-auto']/tr[3]/td[4]/button[2]"
+    DELETE_CATEGORY_XPATH = "//table[@class='table w-100 m-auto']/tr[3]/td[5]/button"
```
Comment (vladskakun) [inline pages/edit_categoty.py:13]: Change "Button" to "Btn" and "Category" to "Ctg"

File: pages/edit_categoty.py (hunk 4 above)
Comment (mehalyna) [reply pages/edit_categoty.py:6]: use multiline docstrings

File: pages/edit_categoty.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+class AdminAddCategoryPage():
+
+    '''Locators and methjods for Admin Edit category page'''
+
+    ADD_CATEGORY_BUTTON_XPATH = "//button[contains(text(),'Add element')]"
+    EDIT_CATEGORY_BUTTON_XPATH = "//table[@class='table w-100 m-auto']/tr[3]/td[4]/div/button"
+    NAME_CATEGORY_INPUT_XPATH = "//table[@class='table w-100 m-auto']/tr[3]/td[1]/input"
+    SUBMIT_EDIT_CATEGORY_XPATH = "//table[@class='table w-100 m-auto']/tr[3]/td[4]/button[@type='submit']"
+    CANCEL_EDIT_CATEGORY_XPATH = "//table[@class='table w-100 m-auto']/tr[3]/td[4]/button[2]"
```
Comment (mehalyna) [inline pages/edit_categoty.py:12]: use naming due to convention

File: pages/edit_categoty.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    DELETE_CATEGORY_XPATH = "//table[@class='table w-100 m-auto']/tr[3]/td[5]/button"
+
```
Comment (mehalyna) [inline pages/edit_categoty.py:14]: methods should each have their own docstrings

File: pages/edit_categoty.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    def find_category_by_name(self, category_name):
+        """
+                Search for a category row by name
+        """
+        category_table = self.driver.find_element(By.CSS_SELECTOR, self.TABLE_OF_CTG_CSS)
+        row_counter = 1
+        for row in category_table.find_elements_by_css_selector('tr > td:nth-child(1)'):
+            row_counter = row_counter + 1
+            if row.text == category_name:
+                self.index_of_row = row_counter
```
Comment (TanyaKalistruk) [inline pages/edit_categoty.py:37]:
```suggestion
        for idx, row in enumerate(category_table.find_elements_by_css_selector('tr > td:nth-child(1)')):
            if row.text == category_name:
                self.index_of_row = idx
```
Probably you need to add 1 to idx, but we can to remove row_counter variable :smile: 
And please add selector to the constant

File: pages/edit_categoty.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    CTG_ADD_SUBMIT_BTN_CSS = "tr:nth-child(1) > td > div > button.text-success"
+    CTG_ADD_CANCEL_BTN_CSS = "tr:nth-child(1) > td > div > button.text-danger"
+    TABLE_OF_CTG_CSS = "table.table.w-100.m-auto"
+    CTG_EDIT_BTN_CSS = f"tr:nth-child({index_of_row}) > td:nth-child(4) > div > button.text-info"
+    CTG_EDIT_NAME_INP_CSS = f"tr:nth-child({index_of_row}) > td:nth-child(1) > form#save-form > input"
+    CTG_SUBMIT_EDIT_BTN_CSS = f"tr:nth-child({index_of_row}) > td:nth-child(4) > div > button.text-success"
+    CTG_CANCEL_EDIT_BTN_CSS = f"tr:nth-child({index_of_row}) > td:nth-child(4) > div > button.text-danger"
+    CTG_DELETE_BTN_CSS = f"tr:nth-child({index_of_row}) > td:nth-child(5) > button"
+    NUMBER_OF_USERS_VALUE_CSS = f"tr:nth-child({index_of_row}) > td:nth-child(2)"
+    NUMBER_OF_EVENTS_VALUE_CSS = f"tr:nth-child({index_of_row}) > td:nth-child(3)"
```
Comment (TanyaKalistruk) [inline pages/edit_categoty.py:21]: Let's make these selectors without f before quotes

File: pages/edit_categoty.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        """
+        category_table = self.driver.find_element(By.CSS_SELECTOR, self.TABLE_OF_CTG_CSS)
+        row_counter = 1
+        for row in category_table.find_elements_by_css_selector('tr > td:nth-child(1)'):
+            row_counter = row_counter + 1
+            if row.text == category_name:
+                self.index_of_row = row_counter
+
+
+    def click_add_category_button(self, locator_of_element):
```
Comment (TanyaKalistruk) [inline pages/edit_categoty.py:40]: But don't you have locator_of_element already specified in the class?

File: pages/edit_categoty.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    CTG_EDIT_NAME_INP_CSS = f"tr:nth-child({index_of_row}) > td:nth-child(1) > form#save-form > input"
+    CTG_SUBMIT_EDIT_BTN_CSS = f"tr:nth-child({index_of_row}) > td:nth-child(4) > div > button.text-success"
+    CTG_CANCEL_EDIT_BTN_CSS = f"tr:nth-child({index_of_row}) > td:nth-child(4) > div > button.text-danger"
+    CTG_DELETE_BTN_CSS = f"tr:nth-child({index_of_row}) > td:nth-child(5) > button"
+    NUMBER_OF_USERS_VALUE_CSS = f"tr:nth-child({index_of_row}) > td:nth-child(2)"
+    NUMBER_OF_EVENTS_VALUE_CSS = f"tr:nth-child({index_of_row}) > td:nth-child(3)"
+
+
+    def __init__(self):
+        self.driver = webdriver.Chrome('C:/Driver/chromedriver.exe')
```
Comment (TanyaKalistruk) [inline pages/edit_categoty.py:25]: Please remove this pass

File: pages/edit_categoty.py [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+    NUMBER_OF_USERS_VALUE_CSS = "tr:nth-child({index_of_row}) > td:nth-child(2)"
+    NUMBER_OF_EVENTS_VALUE_CSS = "tr:nth-child({index_of_row}) > td:nth-child(3)"
+    COLUMN_OF_TABLE_CSS = "tbody > tr"
+
+    def __init__(self):
+        self.driver = webdriver.Chrome()
+
+    def find_category_by_name(self, category_name):
+        """
+                Search for a category row by name
```
Comment (nvoloshchuk19) [inline pages/edit_categoty.py:29]: add, please param `category_name` to docstring

File: pages/edit_categoty.py [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+        """
+        for idx, row in enumerate(self.TABLE_OF_CTG_CSS):
+            if row.text == category_name:
+                self.index_of_row = idx
+
+    def add_ctg_click_btn(self):
```
Comment (nvoloshchuk19) [inline pages/edit_categoty.py:35]: click_add_ctg_btn ?

File: pages/edit_categoty.py [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+        """
+               Method for creating new category with pressing button
+        """
+        element = self.driver.find_element(By.CSS_SELECTOR, self.CTG_ADD_BTN_CSS)
+        element.click()
+
+    def new_ctg_name_inp_into_field(self, data_for_input):
```
Comment (nvoloshchuk19) [inline pages/edit_categoty.py:42]: `enter_ctg_name` ?

File: pages/edit_categoty.py [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+        """
+              Method for entering new category name.
+        :param data_for_input: Variable data_for_input should contain text which
+              we need to enter (name of new category)
+        """
+        element = self.driver.find_element(By.CSS_SELECTOR, self.CTG_ADD_INP_FIELD_CSS)
+        element.send_keys(data_for_input)
+
+    def ctg_cancel_creating(self):
```
Comment (nvoloshchuk19) [inline pages/edit_categoty.py:51]: `click_cancel_ctg_creating_btn`?

File: pages/edit_categoty.py [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+        """
+             Method for canceling of creating new category by pressing CANCEL button
+        """
+        element = self.driver.find_element(By.CSS_SELECTOR, self.CTG_ADD_CANCEL_BTN_CSS)
+        element.click()
+
+    def ctg_submit_creating(self):
```
Comment (nvoloshchuk19) [inline pages/edit_categoty.py:58]: `click_submit_ctg_creating_btn`

File: pages/edit_categoty.py [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+        """
+            Method for pressing submit button to save new category
+        """
+        element = self.driver.find_element(By.CSS_SELECTOR, self.CTG_ADD_SUBMIT_BTN_CSS)
+        element.click()
+
+    def ctg_edit_click_btn(self):
```
Comment (nvoloshchuk19) [inline pages/edit_categoty.py:65]: `click_edit_ctg_btn`?

File: pages/edit_categoty.py [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+        """
+            Method for pressing edit category button
+        """
+        element = self.driver.find_element(By.CSS_SELECTOR, self.CTG_EDIT_BTN_CSS.format(index_of_row = self.index_of_row))
+        element.click()
+
+    def ctg_edit_name_inp(self, input_data):
```
Comment (nvoloshchuk19) [inline pages/edit_categoty.py:72]: `edit_ctg_name`?

File: pages/edit_categoty.py [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+
+
+class AdminAddCategoryPage:
+    """
+        Locators and methods for Admin Edit category page
+    """
+    index_of_row = 0
+    CTG_ADD_BTN_CSS = "button.ml-0"
+    CTG_ADD_INP_FIELD_CSS = "#save-form > div > div > div > input"
+    CTG_ADD_SUBMIT_BTN_CSS = "tr:nth-child(1) > td > div > button.text-success"
```
Comment (nvoloshchuk19) [inline pages/edit_categoty.py:12]: probably it's possible to use one selector for `CTG_ADD_SUBMIT_BTN_CSS ` and `CTG_SUBMIT_EDIT_BTN_CSS `

File: pages/edit_categoty.py [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+    def ctg_edit_name_inp(self, input_data):
+        """
+            Method for clearing input field and entering new name of category
+        :param input_data: new name of category
+        """
+        element = self.driver.find_element(By.CSS_SELECTOR, self.CTG_EDIT_NAME_INP_CSS.format(index_of_row = self.index_of_row))
+        element.clear()
+        element.send_keys(input_data)
+
+    def ctg_edit_submit_btn(self):
```
Comment (nvoloshchuk19) [inline pages/edit_categoty.py:81]: `click_submit_ctg_edit_btn` ?

File: pages/edit_categoty.py [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+        """
+           Method for saving changes in existing category
```
Comment (nvoloshchuk19) [inline pages/edit_categoty.py:83]: `Method for saving change of category name` ?

File: pages/edit_categoty.py [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+        """
+        element = self.driver.find_element(By.CSS_SELECTOR, self.CTG_SUBMIT_EDIT_BTN_CSS.format(index_of_row = self.index_of_row))
+        element.click()
+
+    def ctg_edit_cancel_btn(self):
```
Comment (nvoloshchuk19) [inline pages/edit_categoty.py:88]: `click_cancel_ctg_edit_btn`?

File: pages/edit_categoty.py [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+        """
+            Method for canceling changes
```
Comment (nvoloshchuk19) [inline pages/edit_categoty.py:90]: `Method for saving change of category name`?

File: pages/edit_categoty.py [hunk 24]
```diff
@@ -0,0 +1,102 @@
...
+        """
+        element = self.driver.find_element(By.CSS_SELECTOR, self.CTG_CANCEL_EDIT_BTN_CSS.format(index_of_row=self.index_of_row))
+        element.click()
+
+    def ctg_delete_btn(self):
```
Comment (nvoloshchuk19) [inline pages/edit_categoty.py:95]: `click_delete_ctg_btn`?

**Comment by nvoloshchuk19** [comment]
All my comments it's just suggestions. You can use your own variants. But please take a look at these moments

**Comment by Olya27** [comment]
> All my comments it's just suggestions. You can use your own variants. But please take a look at these moments

I will try to process the comments as soon as possible, thank you very much for all  comments)

**Review by TanyaKalistruk — APPROVED** [review]

File: pages/edit_categoty.py [hunk 25]
```diff
@@ -0,0 +1,109 @@
...
+    CTG_ADD_CANCEL_BTN_CSS = "tr:nth-child(1) button.text-danger"
+    TABLE_OF_CTG_CSS = "table"
+    CTG_EDIT_BTN_CSS = "tr:nth-child({index_of_row})  button.text-info"
+    CTG_EDIT_NAME_INP_CSS = "tr:nth-child({index_of_row}) > td:nth-child(1)  input"
+    CTG_SUBMIT_EDIT_BTN_CSS = "tr:nth-child({index_of_row}) > td:nth-child(4)  button.text-success"
+    CTG_CANCEL_EDIT_BTN_CSS = "tr:nth-child({index_of_row}) > td:nth-child(4)  button.text-danger"
+    CTG_DELETE_BTN_CSS = "tr:nth-child({index_of_row}) > td:nth-child(5)  button"
+    NUMBER_OF_USERS_VALUE_CSS = "tr:nth-child({index_of_row}) > td:nth-child(2)"
+    NUMBER_OF_EVENTS_VALUE_CSS = "tr:nth-child({index_of_row}) > td:nth-child(3)"
+    COLUMN_OF_TABLE_CSS = "tr > td:nth-child(1):not(td.align-middle)"
```
Comment (nvoloshchuk19) [inline pages/edit_categoty.py:24]: It still looks bad, but I hope you will fix it in the near future

**Review by nvoloshchuk19 — APPROVED** [review]

File: pages/edit_categoty.py (hunk 25 above)
Comment (Olya27) [reply pages/edit_categoty.py]: Done, but locator for the Delete button has an identical class to the Cancel button. I did not find a way to get rid of td in Delete button selector. I hope it's better now)

File: pages/edit_categoty.py [hunk 26]
```diff
@@ -0,0 +1,78 @@
...
+    NUMBER_OF_EVENTS_VALUE_CSS = "td:nth-child(3)"
+    COLUMN_OF_TABLE_CSS = "tr > td:nth-child(1):not(td.align-middle)"
+    CTG_ROW = "tr:nth-child({})"
+
+    def __init__(self):
+        self.driver = webdriver.Chrome()
+        self.add_ctg_btn = ButtonElement(self.CTG_ADD_BTN_CSS)
+        self.submit_btn = ButtonElement(self.CTG_SUBMIT_BTN_CSS)
+
+    def find_category_by_name(self, category_name, wait_time=10):
```
Comment (TanyaKalistruk) [inline pages/edit_categoty.py:30]: get_row_idx_by_category_name

**Pull request merged**